Bluetooth Server

Daeda88 opened the pull request:
A significant rewrite of the Bluetooth API (2.0) that focusses primarily on supporting Bluetooth Servers.

### Changes to Client Side
To make the API as good as possible, some breaking changes had to be made:

- `Service`, `Characteristic` and `Descriptor` have been renamed to `RemoteService`, `RemoteCharacteristic` and `RemoteDescriptor` respectively. Their old names have now been reserved for an interface that spans both Server (Local) and Client (Remote) implementations.
- Similarly Device is now named `ConnectableDevice`. From the server side it is called `ConnectedDevice`.
- `BaseBluetoothBuilder.create` has been deprecated. Use `createClient` instead.
- Getting an Attribute from a Flow/List of Attributes by UUID now fails if the UUID is not present. This means that `device.services()[uuid]` may now fail if called before services have been discovered. Use `device.discoveredServices()[uuid]` instead if you only want a flow of the service after discovery (this will fail if service is not present), or use `devices.services.getOrNull(uuid)`. This aligns better with other Kotlin APIs
- `RemoteCharacteristic` and `RemoteDescriptor` are no longer Flows of ByteArray. Instead:
    - ReadAction will result in a `GattResponse.ReadResponse`. If this is `ReadSuccess` the data read will be available as its value.
    - WriteAction will simply return `GattResponse.WriteResponse`.
    - Notifications can be enabled by calling `RemoteCharacteristic.subscribe {}`. The Resulting Subscription must be used to unsubscribe. Calling this method will automatically call `enabledNotifications/disableNotifications()` if needed.
    - `Flow<RemoteCharacteristic?>.value()` will automatically subscribe/unsubscribe as the flow is being collected.
    - `Flow<RemoteDescriptor?>.value()` has been removed.
- The new `Bind` api makes it easy to bind an object to a `ConnectedDevice`, `RemoteService`, `RemoteCharacteristic`, or `RemoteDescriptor`. Simply call `object.bind(service) {}` Within its closure you can set up observation actions and read or write triggers. The resulting StateFlow will help keep track of any mutations that happened to the object.

### Server
- `BaseBluetoothBuilder` has a new method `createServer` to set up a Bluetooth server.
- The resulting server suspends until Any services and Advertisement set up in its DSL has completed.
- Close the resulting server once you are dont to stop it.
- Within the DSL, set up read/write actions to determine the response to a Read/Write request.
- Within the DSL, make a Characteristic Notifiable by calling `notifiable`. The resulting `LocalCharacteristic.Notifiable` has methods for sending notifications to devices. A convenience `collectAsNotification`/`consumeAsNotification` api offers the ability to bind flows directly to a characteristic, notifying each subscriber as the flow updates.
- The BluetoothServer will automatically reboot if Bluetooth is reenabled or permissions are reset. It will restore to the last desired Services and Advertisement.

Daeda88 committed:
WIP Bluetooth server

Daeda88 committed:
API & Android implementation

Daeda88 committed:
iOS Implementation

Daeda88 committed:
Compilable

Daeda88 committed:
Server working with Lifecycle

Daeda88 committed:
iOS State Refactor

Daeda88 committed:
Improvements to Notification behaviour

Daeda88 committed:
Notifiable compiling

Daeda88 committed:
Fixing tests

Daeda88 committed:
Refactoring to move most code to commonMain

Daeda88 committed:
Working on Example

Daeda88 committed:
More Example progress + convenience setup

Byte Utils
Client side binding

Daeda88 committed:
Strip out ByteState from Characteristic and Descriptor

Read/Write will now just provide an accessor for completion.

value() for Descriptor removed
value() for Characteristic will automatically handle subscription

Daeda88 committed:
Example Android Layout

Daeda88 committed:
Improved Error handling

Daeda88 committed:
Convenience methods and improved bind api

Daeda88 committed:
iOS Example

Daeda88 committed:
Example Client working

Read/write still untested

Daeda88 committed:
Server Notifying working (Android)

Daeda88 committed:
Read Write working

Daeda88 committed:
Fixes for closing server

Daeda88 committed:
Fixes for Advertising

Daeda88 committed:
Bluetooth serialization

Daeda88 committed:
Fixed signed error

Daeda88 committed:
API Dump

ci-splendo commented:
### Code coverage
|Total Project Coverage|36.02%|
|:-|:-:|

Daeda88 committed:
Move GattResponse to root Bluetooth package

Daeda88 committed:
Heart Rate encoding working

Daeda88 committed:
Cleaner encoding Architecture

Daeda88 committed:
Enum support

Daeda88 committed:
Collection support

Daeda88 committed:
Sealed classes

Daeda88 committed:
Full Polymorphic support

Daeda88 committed:
Store MedFloat16 as Double for higher precision

Daeda88 committed:
Improved Collection logic

Daeda88 committed:
Fixes for Numeric Encoding

Now properly accounting for Unsigned data

Daeda88 committed:
More tests

Daeda88 committed:
Inline tests

Daeda88 committed:
String tests

Daeda88 committed:
List Tests

Daeda88 committed:
Map tests

Daeda88 committed:
String Decoding

Daeda88 committed:
Numeric decoding fixes

Daeda88 committed:
Decoding working for primitive types

Daeda88 committed:
Deserialization mostly working

Some collection fixes to be done

Daeda88 committed:
Collection Deserialization

Now full support!

Daeda88 committed:
Reenable parts of tests + nicer HR test (real world scenario)

Daeda88 committed:
Convenience methods to Bluetooth Characteristics and Descriptors

Daeda88 committed:
With Example

Daeda88 committed:
(Inline) accessors, Lint & API

Besides Checksum logic, feature complete

Daeda88 committed:
More accessors and some fixes

Daeda88 committed:
CRC Checksum support

thoutbeckers commented on base/src/commonMain/kotlin/utils/ByteArrayBuilder.kt:
```diff
@@ -0,0 +1,127 @@
+
+package com.splendo.kaluga.base.utils
+
+import kotlin.experimental.or
```
Jetbrains please 🤦‍♂️

Daeda88 committed:
Documentation and some cleanup for Base module

thoutbeckers commented on base/src/commonMain/kotlin/utils/ByteArrayBuilder.kt:
```diff
@@ -0,0 +1,127 @@
+fun buildByteArray(order: ByteOrder = ByteOrder.LEAST_SIGNIFICANT_FIRST, block: ByteArrayBuilder.() -> Unit) = ByteArrayBuilderImpl(order).apply(block).bytes
+
+private class ByteArrayBuilderImpl(val order: ByteOrder) : ByteArrayBuilder {
+    var bytes = byteArrayOf()
```
This implementation does have some inefficiencies with copying this byte array over and over (both copies in general but in particular object allocation)

I appreciate however that it's behind an interface, and it's probably the best way to create a reference implementation and validate the test suite etc.

The common optimization is to expand the array in chunks, and have a size suggestion for the first chunk.  Compare API/implementation of https://docs.oracle.com/javase/8/docs/api/java/io/ByteArrayOutputStream.html. In the implementation`currentByte` would be an index pointer in this case, and when you actually write you check the index range to see if you need to add another chunck (I think it's common to double)

Also, converting other values to ByteArray (like Long.toByteArray) could fairly easily be modified to (also) support writing into an existing an existing ByteArray, again avoiding copies.

If the user size hints the chunk correctly (often in such scenarions it's easy to know) and you take care to only expand the array when there is an actual mutation past the boundry you can construct the byte array with 0 copies. In application code you can add a warning if your eventual size does not match your hint so you can build this robustly too.

Alternetly/additionally, a secondary fixed size builder could be considered (if the size is known ahead of time you might as well fail if it's not that size).

To be clear, I do support keeping this implementation in any case. Whether we add a (usually) more optimal one can be up for debate. It seems a bit nitty but even for small byte structures it's easily an order of magnitude less objects. 

I think with the serialization stuff we probably can calculate a good "hint" or even know when the result is fixed size.

thoutbeckers commented on base/src/commonMain/kotlin/utils/ByteExtensions.kt:
```diff
@@ -0,0 +1,23 @@
+package com.splendo.kaluga.base.utils
+
+infix fun Byte.shl(bitCount: Int) = toUByte().shl(bitCount).toByte()
```
AFAIK these methods are free (no runtime cost), but have you considered just going `UByteBuilder` by default?

Daeda88 committed:
More readable code for Encoding/Decoding helper

Daeda88 committed:
Documentation & optimization for Serialization parts

Daeda88 committed:
Better subscription + documentation on root BT + most client stuff

Daeda88 committed:
Documenting Bind

Daeda88 committed:
Server documentation

Daeda88 committed:
Readme

Daeda88 committed:
Merge remote-tracking branch 'origin/develop' into feature/bluetooth-serializer

Daeda88 committed:
Improved Binding api for Mapping Channels/Flows

Daeda88 committed:
Same optimization for LocalCharacteristic

Daeda88 committed:
Fixed Readme

Daeda88 committed:
Remove annoying enable notification  suspension

Daeda88 committed:
Proper notification fix

Daeda88 committed:
Small optimization for atomicity

Daeda88 committed:
Added support for fully customizing BT logging

thoutbeckers commented on bluetooth/build.gradle.kts:
```diff
@@ -4,6 +4,10 @@ plugins {
 }
 
+kotlin {
+    compilerOptions.freeCompilerArgs.add("-Xwhen-guards")
```
is this one we should apply more widely? fairly easy to lift to the plugin level I think.

Daeda88 replied:
Removed, as it is enabled by default in Kotin 2.3

thoutbeckers commented on example/shared/src/commonMain/kotlin/com/splendo/kaluga/example/shared/di/DependencyInjection.kt:
```diff
@@ -43,7 +43,7 @@ typealias BluetoothBuilderBuilder = (suspend (CoroutineContext) -> Permissions)
     Switch this value to use the location permission on Android when using bluetooth.
  */
-const val useBluetoothForLocation = false
+const val USE_LOCATION_FOR_BLUETOOTH = true
```
I think this should be false by default.

Daeda88 replied:
Reverted, but note that since the Example also demoes beacons, which do need the location permission, we cannot add [neverForLocation](https://developer.android.com/develop/connectivity/bluetooth/bt-permissions#assert-never-for-location) and the example will just not scan anything.

thoutbeckers replied:
We don't actually derive location from the beacons.

thoutbeckers replied:
Not having `neverForLocation` is weird though, the example seemed to work without this. Maybe missed commiting it on my side.

thoutbeckers commented on bluetooth/src/androidMain/kotlin/BluetoothBuilder.kt:
```diff
@@ -20,26 +30,43 @@ import kotlin.coroutines.CoroutineContext
         Permissions(
             PermissionsBuilder(PermissionContext(applicationContext)).apply {
                 registerBluetoothPermissionIfNotRegistered()
-                if (useLocationPermission) {
```
Did you retest in both modes (use location and not use location) that the correct permissions are prompted?

(the fact you changed the flag suggests so of course, but best to make sure)

Daeda88 replied:
This code purely *registers* the ability to monitor a Location permission. It does not actually monitor the permissions. Since at this point in time, we wont know the precise permission the user might request, we just register it so its available in case we need it. This registering is a side effect of having multiple permission modules.

thoutbeckers commented on bluetooth/src/commonMain/kotlin/Characteristic.kt:
```diff
@@ -65,23 +85,51 @@ open class Characteristic(
+        subscriptions.forEach { it.onUpdate(value) }
+    }
+
+    private suspend fun unsubscribe(subscription: Subscription): DeviceAction.Notification? = if (subscriptions.remove(subscription) && subscriptions.isEmpty()) {
```
Pontential concurrency problem here, as there is no locking from removing the entry to checking the map. I believe we have the `synchronized` method for this.

thoutbeckers commented on bluetooth/src/commonMain/kotlin/Characteristic.kt:
```diff
@@ -93,140 +141,161 @@ open class Characteristic(
+    constructor(rawValue: Int) : this(rawValue, rawValue)
+
+    companion object {
+        fun fromInt(properties: Int): Set<CharacteristicProperty> = setOf(
```
seems like we don't need to recreate this set every time?

thoutbeckers commented on bluetooth/src/commonMain/kotlin/device/Bind.kt:
```diff
@@ -0,0 +1,995 @@
+import kotlinx.coroutines.launch
+import kotlin.jvm.JvmName
+
+sealed interface RequiresServicesDiscoveredBinding<T, TransFormation> {
```
nit:
```suggestion
sealed interface RequiresServicesDiscoveredBinding<T, Transformation> {
```

thoutbeckers commented on test-utils-bluetooth/src/commonMain/kotlin/device/MockDeviceConnectionManager.kt:
```diff
@@ -164,58 +168,63 @@ class MockDeviceConnectionManager(
+                action = it
+                true
+            } ?: run {
+                delay(100)
```
Even though this is just mocking, it seems like we can solve this better, in a suspend context especially?

thoutbeckers commented on bluetooth/src/commonMain/kotlin/server/BluetoothServer.kt:
```diff
@@ -0,0 +1,594 @@
+                        } else {
+                            logger.warn(TAG) { "Failed to add service ${service.uuid}" }
+                            serviceAction.isAdded.complete(null)
+                        }
```
There seem to be no suspend calls in the `NonCancellable` blocks, so it's technically not needed.

thoutbeckers commented on bluetooth/src/commonMain/kotlin/server/LocalCharacteristic.kt:
```diff
@@ -0,0 +1,374 @@
+            encrypted: Boolean = false,
+            toByteArray: T.() -> ByteArray,
+        ) {
+            val hasStarted = CompletableDeferred<Unit>()
```
Kaluga has `EmptyCompletableDeferred` since we use these quite often.

thoutbeckers commented on bluetooth/src/commonTest/kotlin/BluetoothCharacteristicValueTest.kt:
```diff
@@ -53,16 +57,21 @@ class BluetoothCharacteristicValueTest : BluetoothFlowTest<BluetoothFlowTest.Con
+            val captor = AnyOrNullCaptor<DeviceAction<*>>()
+            connectionManager.performActionMock.verifyWithin(value = captor)
+            assertIs<DeviceAction.Notification.Enable>(captor.lastCaptured)
+            yield()
```
the cause of having to do these random `yields` is probably still using `BluetoothFlowTest` and `mainAction`, something from the times of not having background dispatchers on iOS.

thoutbeckers commented on bluetooth/src/commonTest/kotlin/BluetoothDistanceTest.kt:
```diff
@@ -38,6 +38,9 @@ class BluetoothDistanceTest : BluetoothFlowTest<BluetoothFlowTest.Configuration.
         ),
     ) {
+        test {
+            assertEquals(Double.NaN, it)
```
Not the most used API but could probably use a (value) type in a 2.0 API.

thoutbeckers commented on bluetooth/src/commonMain/kotlin/server/BluetoothServer.kt:
```diff
@@ -0,0 +1,594 @@
+    private val logger = settings.logger
+    private val _status = MutableStateFlow(initialState.status)
+    val status: StateFlow<ServerStatus> = _status.asStateFlow()
+    private val _isAdvertising = MutableStateFlow(false)
```
more of a consideration, Kotlin 2.3 has [explicit backing fields](https://kotlinlang.org/docs/whatsnew23.html#explicit-backing-fields) for this now (needs to be enabled)

Daeda88 replied:
Read Only modifying backing fields are not supported though: https://github.com/Kotlin/KEEP/blob/explicit-backing-fields/proposals/explicit-backing-fields.md#returning-read-only-view

thoutbeckers commented on bluetooth/src/iosMain/kotlin/server/BluetoothServer.kt:
```diff
@@ -0,0 +1,198 @@
+    ) : IOSServerState(),
+        ServerState.AwaitingBluetoothEnabled {
+
+        private val serverQueue = dispatch_queue_create("BluetoothServer", null)
```
Should this not be raised up, maybe as a lazy field? AFAIK these queues will not autoclose.

Daeda88 replied:
They should auto close. There is also no way of closing them manually as `dispatch_release` has been deprecated and Apple explicitly says not to use it (anymore)

thoutbeckers commented on bluetooth/src/iosMain/kotlin/server/KalugaCBPeripheralManagerDelegate.kt:
```diff
@@ -0,0 +1,189 @@
+        return available
+    }
+
+    private val readActions = mutableMapOf<CBCharacteristic, suspend (ConnectedDevice, Int) -> GattResponse.ReadResponse>()
+    private val writeActions = mutableMapOf<CBCharacteristic, suspend (ConnectedDevice, ByteArray, Int) -> GattResponse.WriteResponse>()
+    private val subscribeActions = mutableMapOf<CBCharacteristic, (ConnectedDevice) -> Unit>()
+    private val unsubscribeActions = mutableMapOf<CBCharacteristic, (ConnectedDevice) -> Unit>()
+
+    private val handlingScope = CoroutineScope(handlingContext + CoroutineName("CBPeripheralManagerDelegate"))
```
Are the these accessed from just one thread?

thoutbeckers reviewed (commented):
some small things that can be looked at, other things (efficiency etc) could be for later consideration.

I do also think more extensive tests is of course desirable, but it's worth it to wait for server <-> client mocking perhaps.

Daeda88 committed:
Merge remote-tracking branch 'origin/develop' into feature/bluetooth-server

Daeda88 committed:
Merge branch 'feature/bluetooth-server' into feature/bluetooth-serializer

Daeda88 committed:
Example formatting & Import optimizations

Daeda88 committed:
Merge branch 'feature/bluetooth-server' into feature/bluetooth-serializer

Daeda88 committed:
Merge branch 'feature/bluetooth-serializer' into feature/bluetooth-finegrained-logging

Daeda88 committed:
Process several PR remarks

Daeda88 committed:
Refactor ByteArrayBuilder to be more efficient

Less copying used in implementation

Daeda88 commented:
@thoutbeckers the most pressing issues have been fixed here https://github.com/splendo/kaluga/pull/925

Daeda88 committed:
Optimizations + disable explicit backing fields for now

Daeda88 committed:
Fixed range check

Daeda88 committed:
Merge pull request #925 from splendo/feature/bluetooth-pr-remarks

PR remarks for BT

Daeda88 committed:
Merge pull request #918 from splendo/feature/bluetooth-finegrained-logging

Added support for fully customizing BT logging

Daeda88 committed:
Merge pull request #914 from splendo/feature/bluetooth-serializer

Bluetooth Serializer

Daeda88 committed:
Adding even better support for Encoding

Daeda88 committed:
Copy safe String behaviour

Daeda88 committed:
Further optimizations

Daeda88 committed:
Infer ByteArray size on encoding

Daeda88 committed:
Make extension accessors have same safety as other functions

Daeda88 committed:
Merge pull request #928 from splendo/feature/bt-accessors

Make extension accessors have same safety as other functions

Daeda88 committed:
Merge remote-tracking branch 'origin/develop' into feature/byte_array_builder_encoding_extensions

Daeda88 committed:
API

Daeda88 committed:
UTF8 encoding with malformed support

Daeda88 committed:
Documentation

Daeda88 committed:
Code optimization

Daeda88 committed:
Merge pull request #926 from splendo/feature/byte_array_builder_encoding_extensions

More ByteBuilder optimizations

Daeda88 committed:
Fix for isNotifying not being a StateFlow

Daeda88 committed:
Merge pull request #932 from splendo/feature/bt_notifying_status

Fix for isNotifying not being a StateFlow

Daeda88 committed:
Reintroducing Weak references fix

Daeda88 committed:
Server strong reference fix

Daeda88 committed:
ABI

Daeda88 committed:
Merge pull request #934 from splendo/feature/weak-reference-fix-revert

Weak reference fix redux

thoutbeckers committed:
Bump Kotlin to 2.3.20-RC3

thoutbeckers committed:
Replace deprecated excluded() with exclude() for Kotlin 2.3.20-RC3

thoutbeckers committed:
Bump Kotlin to 2.3.20 final

thoutbeckers committed:
Fix rerun_emulator workflow missing actions: write permission

thoutbeckers committed:
Migrate to Gradle 9.4.0, AGP 9.1.0, Kotlin 2.2.0

- Bump Gradle wrapper to 9.4.0
- Bump AGP to 9.1.0-beta01 and related dependencies
- Add IgnoreJs annotation to test-utils-base for platform-specific test skipping
- Skip JS tests that rely on synchronous runBlocking (UIThreadViewModel,
  TestBlockingAndCancelScope, UIThreadTest exception propagation, Koin)
- Skip JS tests with Node.js ICU/locale limitations (DateFormatter,
  NumberFormatter, StringFormatter, DateTest, DecimalTest, LocaleTest)
- Skip JS tests with BigDecimal precision differences (scientific,
  scientific-converters unit and converter tests)
- Fix ExperimentalMaterialApi opt-in for AGP 9.0 built-in Kotlin:
  use compilerOptions.optIn instead of languageSettings
- Restore androidDeviceTest dependsOn(commonTest) with explanatory comment

thoutbeckers committed:
Update publish workflow to JDK 21

thoutbeckers committed:
Bump JVM args for daemon reuse with health-app

thoutbeckers committed:
Update CI to JDK 21

thoutbeckers committed:
Add foojay toolchain resolver, daemon JVM config, and catalog entries

thoutbeckers committed:
Reduce Gradle heap from 13g to 10g

thoutbeckers committed:
Upgrade example to Gradle 9.4.1 + AGP 9 with legacy-kapt for databinding

thoutbeckers committed:
Fix CI for Gradle 9: iOS sim OS 26.3.1, failOnNoDiscoveredTests, test-utils-base ABI

thoutbeckers committed:
Use iOS sim OS 26.4 (available on runner; 26.3.1 not provisioned)

thoutbeckers committed:
Add JS tests workflow

thoutbeckers committed:
Align main wrapper with example: Gradle 9.4.1

thoutbeckers committed:
Modernize gradle.properties JVM args

Xmx 10g→12g and Xms 256m→4g to cut heap-growth churn and allow
memory-heavy K/N link tasks. Drop -XX:+UseParallelGC (G1 is the
JDK 21 default and better for long-lived daemons) and drop
-Dfile.encoding=UTF-8 (default in JDK 18+). Add
-XX:ReservedCodeCacheSize=512m to avoid code-cache exhaustion
on long-lived daemons.

thoutbeckers committed:
CI: iOS sim OS 26.4 → 26.4.1 (runner image refreshed)

thoutbeckers committed:
Bump Kotlin 2.3.20 → 2.3.21

thoutbeckers committed:
CI: pin iOS sim destination to OS=latest

Runner image rotates between patch versions (26.4, 26.4.1, ...), so any
exact OS string breaks intermittently. 'latest' resolves to whichever
matching iPhone 17 runtime the runner has.

thoutbeckers committed:
Fix flaky NetworkStateTest.testUnknownTransition

After resetFlow(), the new collector subscribes to NetworkStateRepo's
cached state (Cellular at this point) before the subsequent value
change has propagated. Sometimes that replay arrived first, causing
the next test{} block to see Cellular instead of the expected Wifi.

Consume the Cellular replay explicitly so the race doesn't matter.

thoutbeckers committed:
Bump kotlinx-kover 0.9.4 → 0.9.8

thoutbeckers committed:
Merge pull request #937 from splendo/feature/kotlin-2.3.20

Kotlin 2.3.20 upgrade

thoutbeckers committed:
Merge pull request #938 from splendo/feature/gradle-9

Gradle 9.x, AGP 9.x

thoutbeckers committed:
Add regression tests for byte and Bluetooth serialization bugs

thoutbeckers committed:
Fix byte-encoding and Bluetooth serialization bugs
